conan/tools/system/package_manager.py

```diff
@@ -0,0 +1,208 @@
+from conans.errors import ConanException
+
+
+class SystemPackageManagerTool(object):
+    install_methods = ["install", "update"]
+    mode_check = "check"
+    tool_name = None
+    install_command = ""
+    update_command = ""
+    check_command = ""
+
+    def __init__(self, conanfile):
+        self._conanfile = conanfile
+        self._active_tool = self._conanfile.conf["tools.system.package_manager:tool"] or self.get_default_tool()
+        self._sudo = self._conanfile.conf["tools.system.package_manager:sudo"]
+        self._sudo_askpass = self._conanfile.conf["tools.system.package_manager:sudo_askpass"]
+        self._mode = self._conanfile.conf["tools.system.package_manager:mode"] or self.mode_check
+        self._arch = self._conanfile.settings_build.get_safe('arch') \
+            if self._conanfile.context == CONTEXT_BUILD else self._conanfile.settings.get_safe('arch')
+        self._arch_names = {}
+        self._arch_separator = ""
+
+    def get_default_tool(self):
+        os_name = platform.system()
+        if os_name in ["Linux", "FreeBSD"]:
+            import distro
+            os_name = distro.id() or os_name
+        elif os_name == "Windows" and self._conanfile.conf["tools.microsoft.bash:subsystem"] == "msys2":
+            os_name = "msys2"
+        manager_mapping = {"apt-get": ["Linux", "ubuntu", "debian"],
+                           "yum": ["pidora", "scientific", "xenserver", "amazon", "oracle", "amzn",
+                                   "almalinux"],
+                           "dnf": ["fedora", "rhel", "centos", "mageia"],
+                           "brew": ["Darwin"],
+                           "pacman": ["arch", "manjaro", "msys2"],
+                           "choco": ["Windows"],
+                           "zypper": ["opensuse", "sles"],
+                           "pkg": ["freebsd"],
+                           "pkgutil": ["Solaris"]}
+        for tool, distros in manager_mapping.items():
+            if os_name in distros:
+                return tool
+
+    def get_package_name(self, package):
+        # TODO: should we only add the arch if cross-building?
+        if self._arch in self._arch_names:
+            return "{}{}{}".format(package, self._arch_separator,
+                                   self._arch_names.get(self._arch))
+        return package
+
+    @property
+    def sudo_str(self):
+        sudo = "sudo " if self._sudo else ""
+        askpass = "-A " if self._sudo and self._sudo_askpass else ""
+        return "{}{}".format(sudo, askpass)
+
+    def check_enabled_tool(wrapped):
+        def wrapper(self, *args, **kwargs):
+            if self._active_tool == self.__class__.tool_name:
+                return wrapped(self, *args, **kwargs)
+
+        return wrapper
+
+    def check_mode(wrapped):
+        def wrapper(self, *args, **kwargs):
+            method_name = wrapped.__name__
+            if method_name in self.install_methods and self._mode == self.mode_check:
```

Comment:
I am not sure if I fully understand the logic. I think that the tools, even in check mode if possible should check if the requested packages are installed, and if they are, it shouldn't be an error, it should only raise if it really needs to install those packages.

I don't see how this implements this behavior, but maybe I am missing something.

Reply:
I have updated the logic and removed the decorator for the mode check, now:
- In `check` mode:
   - `update` fails
   - `install` will check if missing packages by default (you can deactivate this) and if there are missing packages will fail, if not it will continue
- In `install` mode:
   - `update` succeeds
   - `install` install will check if missing packages by default (you can deactivate this) and if there are missing packages will try to install them